File: ios/FestivalConnection/Protocol/CrowdSyncPacket.swift
import Foundation

// MARK: - CrowdSync Packet
// Binary packet structure copied from BitChat reference for cross-platform compatibility.
// Wire format matches BitChat's BinaryProtocol exactly.

struct CrowdSyncPacket: Codable {
    let version: UInt8
    let type: UInt8
    let senderID: Data       // 8 bytes
    let recipientID: Data?   // 8 bytes (optional, flag bit 0)
    let timestamp: UInt64    // milliseconds since epoch
    let payload: Data
    var signature: Data?     // 64 bytes (optional, flag bit 1)
    var ttl: UInt8

    init(type: UInt8, senderID: Data, recipientID: Data? = nil,
         timestamp: UInt64 = UInt64(Date().timeIntervalSince1970 * 1000),
         payload: Data, signature: Data? = nil, ttl: UInt8 = 7,
         version: UInt8 = 1) {
        self.version = version
        self.type = type
        self.senderID = senderID
        self.recipientID = recipientID
        self.timestamp = timestamp
        self.payload = payload
        self.signature = signature
        self.ttl = ttl
    }
}

// MARK: - Binary Protocol Encoding/Decoding
// Header structure (v1, 14 bytes):
//   [version:1][type:1][ttl:1][timestamp:8][flags:1][payloadLength:2]
// Then: [senderID:8][recipientID?:8][payload:var][signature?:64]

struct CrowdSyncBinaryProtocol {
    static let v1HeaderSize = 14
    static let senderIDSize = 8
    static let recipientIDSize = 8
    static let signatureSize = 64

    struct Flags {
        static let hasRecipient: UInt8  = 0x01
        static let hasSignature: UInt8  = 0x02
        static let isCompressed: UInt8  = 0x04
    }

    static func encode(_ packet: CrowdSyncPacket) -> Data? {
        var data = Data()
        data.reserveCapacity(v1HeaderSize + senderIDSize + packet.payload.count + 100)

        data.append(packet.version)
        data.append(packet.type)
        data.append(packet.ttl)

        // Timestamp: 8 bytes big-endian
        for shift in stride(from: 56, through: 0, by: -8) {
            data.append(UInt8((packet.timestamp >> UInt64(shift)) & 0xFF))
        }

        // Flags
        var flags: UInt8 = 0
        if packet.recipientID != nil { flags |= Flags.hasRecipient }
        if packet.signature != nil  { flags |= Flags.hasSignature }
        data.append(flags)

        // Payload length: 2 bytes big-endian (v1)
        let payloadLength = UInt16(min(packet.payload.count, Int(UInt16.max)))
        data.append(UInt8((payloadLength >> 8) & 0xFF))
        data.append(UInt8(payloadLength & 0xFF))

        // Sender ID (8 bytes, zero-padded)
        let senderBytes = packet.senderID.prefix(senderIDSize)
        data.append(senderBytes)
        if senderBytes.count < senderIDSize {
            data.append(Data(repeating: 0, count: senderIDSize - senderBytes.count))
        }

        // Optional recipient ID
        if let recipientID = packet.recipientID {
            let recipientBytes = recipientID.prefix(recipientIDSize)
            data.append(recipientBytes)
            if recipientBytes.count < recipientIDSize {
                data.append(Data(repeating: 0, count: recipientIDSize - recipientBytes.count))
            }
        }

        // Payload
        data.append(packet.payload)

        // Optional signature
        if let signature = packet.signature {
            data.append(signature.prefix(signatureSize))
        }

        return data
    }

    static func decode(_ data: Data) -> CrowdSyncPacket? {
        guard data.count >= v1HeaderSize + senderIDSize else { return nil }

        var offset = 0
        func read8() -> UInt8? {
            guard offset < data.count else { return nil }
            let v = data[offset]; offset += 1; return v
        }
        func readData(_ n: Int) -> Data? {
            guard offset + n <= data.count else { return nil }
            let d = data[offset..<(offset + n)]; offset += n; return Data(d)
        }

        guard let version = read8(), version == 1 else { return nil }
        guard let type = read8(), let ttl = read8() else { return nil }

        var timestamp: UInt64 = 0
        for _ in 0..<8 {
            guard let byte = read8() else { return nil }
            timestamp = (timestamp << 8) | UInt64(byte)
        }

        guard let flags = read8() else { return nil }
        let hasRecipient = (flags & Flags.hasRecipient) != 0
        let hasSignature = (flags & Flags.hasSignature) != 0

        guard let lenHi = read8(), let lenLo = read8() else { return nil }
        let payloadLength = Int((UInt16(lenHi) << 8) | UInt16(lenLo))

        guard let senderID = readData(senderIDSize) else { return nil }

        var recipientID: Data?
        if hasRecipient {
            recipientID = readData(recipientIDSize)
            if recipientID == nil { return nil }
        }

        guard let payload = readData(payloadLength) else { return nil }

        var signature: Data?
        if hasSignature {
            signature = readData(signatureSize)
            if signature == nil { return nil }
        }

        return CrowdSyncPacket(
            type: type, senderID: senderID, recipientID: recipientID,
            timestamp: timestamp, payload: payload, signature: signature,
            ttl: ttl, version: version
        )
    }
}

File: android/app/src/main/java/com/appxstudios/festivalconnection/protocol/CrowdSyncPacket.kt
package com.appxstudios.festivalconnection.protocol

// Binary packet structure matching iOS CrowdSyncPacket and BitChat wire format.
// Header v1 (14 bytes): [version:1][type:1][ttl:1][timestamp:8][flags:1][payloadLength:2]

data class CrowdSyncPacket(
    val version: Int = 1,
    val type: Int,
    val senderID: ByteArray,
    val recipientID: ByteArray? = null,
    val timestamp: Long = System.currentTimeMillis(),
    val payload: ByteArray,
    val signature: ByteArray? = null,
    var ttl: Int = 7
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CrowdSyncPacket) return false
        return version == other.version && type == other.type && timestamp == other.timestamp &&
            senderID.contentEquals(other.senderID)
    }
    override fun hashCode(): Int = timestamp.hashCode() + type
}

object CrowdSyncBinaryProtocol {
    const val V1_HEADER_SIZE = 14
    const val SENDER_ID_SIZE = 8
    const val RECIPIENT_ID_SIZE = 8
    const val SIGNATURE_SIZE = 64
    const val FLAG_HAS_RECIPIENT = 0x01
    const val FLAG_HAS_SIGNATURE = 0x02
    // Wire-format parity with iOS CrowdSyncBinaryProtocol.Flags.isCompressed
    // (Protocol/CrowdSyncPacket.swift:46). Reserved for forward compatibility —
    // neither platform emits or consumes compressed payloads yet, but receivers
    // need to recognize the flag bit so a future compressed-capable peer doesn't
    // get its packets rejected as malformed.
    const val FLAG_IS_COMPRESSED = 0x04

    fun encode(packet: CrowdSyncPacket): ByteArray? {
        val buf = mutableListOf<Byte>()
        buf.add(packet.version.toByte())
        buf.add(packet.type.toByte())
        buf.add(packet.ttl.toByte())

        for (shift in 56 downTo 0 step 8) {
            buf.add(((packet.timestamp shr shift) and 0xFF).toByte())
        }

        var flags = 0
        if (packet.recipientID != null) flags = flags or FLAG_HAS_RECIPIENT
        if (packet.signature != null) flags = flags or FLAG_HAS_SIGNATURE
        buf.add(flags.toByte())

        val payloadLen = packet.payload.size.coerceAtMost(65535)
        buf.add(((payloadLen shr 8) and 0xFF).toByte())
        buf.add((payloadLen and 0xFF).toByte())

        val sender = packet.senderID.copyOf(SENDER_ID_SIZE)
        buf.addAll(sender.toList())

        packet.recipientID?.let {
            val recipient = it.copyOf(RECIPIENT_ID_SIZE)
            buf.addAll(recipient.toList())
        }

        buf.addAll(packet.payload.toList())

        packet.signature?.let {
            buf.addAll(it.take(SIGNATURE_SIZE).toList())
        }

        return buf.toByteArray()
    }

    fun decode(data: ByteArray): CrowdSyncPacket? {
        if (data.size < V1_HEADER_SIZE + SENDER_ID_SIZE) return null
        var offset = 0

        val version = data[offset++].toInt() and 0xFF
        if (version != 1) return null
        val type = data[offset++].toInt() and 0xFF
        val ttl = data[offset++].toInt() and 0xFF

        var timestamp = 0L
        repeat(8) { timestamp = (timestamp shl 8) or (data[offset++].toLong() and 0xFF) }

        val flags = data[offset++].toInt() and 0xFF
        val hasRecipient = (flags and FLAG_HAS_RECIPIENT) != 0
        val hasSignature = (flags and FLAG_HAS_SIGNATURE) != 0

        val payloadLength = ((data[offset].toInt() and 0xFF) shl 8) or (data[offset + 1].toInt() and 0xFF)
        offset += 2

        if (offset + SENDER_ID_SIZE > data.size) return null
        val senderID = data.sliceArray(offset until offset + SENDER_ID_SIZE)
        offset += SENDER_ID_SIZE

        val recipientID = if (hasRecipient) {
            if (offset + RECIPIENT_ID_SIZE > data.size) return null
            val r = data.sliceArray(offset until offset + RECIPIENT_ID_SIZE)
            offset += RECIPIENT_ID_SIZE
            r
        } else null

        if (offset + payloadLength > data.size) return null
        val payload = data.sliceArray(offset until offset + payloadLength)
        offset += payloadLength

        val signature = if (hasSignature && offset + SIGNATURE_SIZE <= data.size) {
            val s = data.sliceArray(offset until offset + SIGNATURE_SIZE)
            offset += SIGNATURE_SIZE
            s
        } else null

        return CrowdSyncPacket(version, type, senderID, recipientID, timestamp, payload, signature, ttl)
    }
}
